src/Lecture4.hs

```diff
@@ -184,7 +211,39 @@ instance for the 'Stats' type itself.
 -}
 
 instance Semigroup Stats where
-
+  (<>) :: Stats -> Stats -> Stats
+  (<>) s1 s2 =
+    Stats
+      { statsTotalPositions = statsTotalPositions s1 <> statsTotalPositions s2,
+        statsTotalSum = statsTotalSum s1 <> statsTotalSum s2,
+        statsAbsoluteMax = statsAbsoluteMax s1 <> statsAbsoluteMax s2,
+        statsAbsoluteMin = statsAbsoluteMin s1 <> statsAbsoluteMin s2,
+        statsSellMax = statsSellMax s1 <> statsSellMax s2,
+        statsSellMin = statsSellMin s1 <> statsSellMin s2,
+        statsBuyMax = statsBuyMax s1 <> statsBuyMax s2,
+        statsBuyMin = statsBuyMin s1 <> statsBuyMin s2,
+        statsLongest = statsLongest s1 <> statsLongest s2
+      }
+
+statsAppend' :: Stats -> Stats -> Stats
+statsAppend' s1 s2 =
+  Stats
+    { statsTotalPositions = statsTotalPositions s1 <> statsTotalPositions s2,
+      statsTotalSum = statsTotalSum s1 <> statsTotalSum s2,
+      statsAbsoluteMax = statsAbsoluteMax s1 <> statsAbsoluteMax s2,
+      statsAbsoluteMin = statsAbsoluteMin s1 <> statsAbsoluteMin s2,
+      statsSellMax = statsSellMax s1 `maybeAppend'` statsSellMax s2,
+      statsSellMin = statsSellMin s1 `maybeAppend'` statsSellMin s2,
+      statsBuyMax = statsBuyMax s1 `maybeAppend'` statsBuyMax s2,
+      statsBuyMin = statsBuyMin s1 `maybeAppend'` statsBuyMin s2,
+      statsLongest = statsLongest s1 <> statsLongest s2
+    }
+
+maybeAppend' :: Semigroup a => Maybe a -> Maybe a -> Maybe a
+maybeAppend' Nothing Nothing = Nothing
+maybeAppend' (Just !x) Nothing = Just x
+maybeAppend' Nothing (Just !x) = Just x
+maybeAppend' (Just !x) (Just !y) = Just (x <> y)
```

Comment:
That's absolutely correct implementation 👍🏻 
You can write it a bit shorter though by pattern matching only on the final result:

```suggestion
maybeAppend' a b = case a <> b of
    Nothing -> Nothing
    Just !x -> Just x
```

Reply:
Indeed, that's a bit cleaner by thinking the other way around.